gitlab-runner/src/run.rs

```diff
@@ -54,36 +55,74 @@ where
         Ok(())
     };
 
-    let upload = !cancel_token.is_cancelled()
-        && response.artifacts.get(0).map_or(false, |a| match a.when {
-            ArtifactWhen::Always => true,
-            ArtifactWhen::OnSuccess => script_result.is_ok(),
-            ArtifactWhen::OnFailure => script_result.is_err(),
-        });
-
-    let r = if upload {
-        if let Ok(mut uploader) = Uploader::new(client, &build_dir, response) {
-            let r = handler.upload_artifacts(&mut uploader).await;
-            if r.is_ok() {
-                uploader.upload().await.and(script_result)
-            } else {
-                r
+    let mut overall_result = script_result;
+
+    for artifact in response.artifacts.iter().take(1) {
+        let upload = !cancel_token.is_cancelled()
+            && match artifact.when {
+                ArtifactWhen::Always => true,
+                ArtifactWhen::OnSuccess => script_result.is_ok(),
+                ArtifactWhen::OnFailure => script_result.is_err(),
+            };
+        if !upload {
+            continue;
+        }
+        if let Ok(mut uploader) = Uploader::new(client.clone(), &build_dir, response.clone()) {
+            let mut available_files: Vec<U> = handler.get_uploadable_files().await?;
+            let mut res = Ok(());
+            let mut uploaded = 0;
+
+            for path in artifact.paths.iter() {
+                let matches = match glob::Pattern::new(path) {
+                    Ok(pattern) => available_files
```

Comment:
This is may cause files to be matched and uploaded multiple times.. This should really be turned around so for each artifacts it filters if any of the paths match and then uploads

Reply:
@sjoerdsimons I'm not sure I follow here. The `available_files` mess (obviously not elegant) takes care of multiple uploads (we delete things from `available_files` when we upload them for a given artifact). But the fact that the same file can be uploaded multiple times _for_different_artifacts_ is required, this is an important part of matching Gitlab's behaviour. For example, if you specify `junit.xml` in reports and under the normal artifacts, it is uploaded twice. The first time it's going into the zip file (that's one artifact), and the second time it's getting gzipped and bypassing the normal pathways inside Gitlab, so that it enters the web interface.

The actual "upload_multiple_artifacts" feature tag means we can repackage data multiple times in different ways to make Gitlab happy. It describes to us what it wants in each one, but they are isolated from each other.

Reply:
Oh i see you use a mutable vector and then remove the ones you uploaded.. ugh that's hard to parse as a code flow